[im] Land ReadAttribute over Interaction Model protocal

erjiaqing opened the pull request:
#### Problem
* Fixed #6360
* We are not using interaction model for reading cluster attributes before.

#### Change overview
* Use interaction model for reading cluster attributes

#### Testing
- Local tested using chip-device-ctrl
- Protocol shift, already covered by Darwin and Cirque tests

msandstedt commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -404,6 +404,7 @@ CHIP_ERROR DeviceController::GetDevice(NodeId deviceId, const SerializedDevice &
         }
 
         device->Init(GetControllerDeviceInitParams(), mListenPort, mAdminId);
+        ReturnErrorOnFailure(device->LoadSecureSessionParametersIfNeeded());
```
What specifically is this fixing?

erjiaqing replied:
@pan-apple 
I found without this line, run following command may result in MsgCounter verify failure:
```
connect
zclread Cluster Attribute ...
zcl Cluster Command ...
```

After some research, I found the device get by GetDevice is not at a activate state. So I added this to set it to active before sending any commands.

bzbarsky-apple replied:
@pan-apple Could you please look over this part?

msandstedt commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +97,128 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    switch (dataType)
```
This switch statement is pretty hard to follow with all of the fall through cases.

Is it possible to simplify this?  I don't think this type of brevity is always useful.  It doesn't meaningfully reduce code space and makes bugs harder to see.

erjiaqing replied:
Maybe we can introduce some macro / templates here.

I think the code without FALLTHROUGH is more clear, however, the compiler is not happy to see it.

erjiaqing replied:
Added "BaseType" to make the code clearer here.

yunhanw-google commented on src/app/MessageDef/AttributePath.h:
```diff
@@ -113,7 +113,7 @@ class Parser : public chip::app::Parser
      *          #CHIP_END_OF_TLV if there is no such element
      */
-    CHIP_ERROR GetFieldId(chip::FieldId * const apFieldId) const;
+    CHIP_ERROR GetFieldId(chip::AttributeId * const apFieldId) const;
```
rename all FieldId with attributeId?

erjiaqing replied:
Done

woody-apple commented:
/rebase

woody-apple commented:
@erjiaqing Appears to be a real build failure

andy31415 reviewed (approved):
Approving, however noting that due to generated code, I cannot properly review a 8K added,1K deleted PR.

Mentioning this as a signal flag that I am suffering from generated code when doing code reviews.

woody-apple reviewed (changes requested):
Requesting changes based on a conversation with @vivien-apple @bzbarsky-apple 

Note: This makes the bridge app fail (@jrhees-cae):
```
[1623860061821] [0xe80109] CHIP: [IN] Secure transport received message destined to fabric 0, node 0x0000000000BC5C01. Key ID 0
[1623860061821] [0xe80109] CHIP: [EM] Received message of type 16 and protocolId 0 on exchange 44251
[1623860076378] [0xe80109] CHIP: [IN] Secure transport received message destined to fabric 0, node 0x0000000000BC5C01. Key ID 0
[1623860076378] [0xe80109] CHIP: [EM] Received message of type 2 and protocolId 5 on exchange 44249
[1623860076378] [0xe80109] CHIP: [EM] ec id: 44249, Delegate: 0x0x10f4365a0
[1623860076378] [0xe80109] CHIP: [DMG] Receive Read request
[1623860076378] [0xe80109] CHIP: [-] CHIP Error 4003 (0x00000FA3): Incorrect state at ../../src/app/ReadHandler.cpp:47
[1623860076378] [0xe80109] CHIP: [-] CHIP Error 4003 (0x00000FA3): Incorrect state at ../../src/app/InteractionModelEngine.cpp:225
[1623860076378] [0xe80109] CHIP: [EM] Sending Standalone Ack for MsgId:00000006
[1623860076378] [0xe80109] CHIP: [IN] Secure message was encrypted: Msg ID 4
[1623860076378] [0xe80109] CHIP: [IN] Send message of type 16 and protocolId 0 on exchange 44249
[1623860076378] [0xe80109] CHIP: [IN] Sending msg from 0x0000000000BC5C01 to 0x000000000001B669 at utc time: 2244760720 msec
[1623860076378] [0xe80109] CHIP: [IN] Sending secure msg on generic transport
[1623860076379] [0xe80109] CHIP: [IN] Secure msg send status No Error
[1623860076379] [0xe80109] CHIP: [EM] Flushed pending ack for MsgId:00000006
```

Also doesn't support lists yet. We'll need tests in place to verify those as well @vivien-apple (who I know is working on that).

bzbarsky-apple commented:
@erjiaqing For review purposes it would be very helpful if there was one commit in the PR for the non-generated changes and one commit for all the generated bits.  Then reviewers could basically read the non-generated commit and skim over the generated results to make sure nothing surprising happens.  right now reviewing this involves reviewing either a bunch of not-actually-meant-to-be-working commits or manually trying to separate out the generated changes from the rest of the PR diff....

yunhanw-google commented:
Hi, @woody-apple @bzbarsky-apple 
we will fix bridge app and rebase this PR with rebase -i, it seems we need manually validate it with bridge app?  is that only concern for this PR? 

for list support, since @vivien-apple is working on it, I would assume we don't need to worry about the list support in this PR?

woody-apple commented:
/rebase

woody-apple commented:
> Hi, @woody-apple @bzbarsky-apple
> we will fix bridge app and rebase this PR with rebase -i, it seems we need manually validate it with bridge app? is that only concern for this PR?
> 
> for list support, since @vivien-apple is working on it, I would assume we don't need to worry about the list support in this PR?

Deferring to @vivien-apple, but this breaks existing functionality via manual testing. We should make sure we have tests in place to verify existing functionality via CI.

bzbarsky-apple commented:
> I would assume we don't need to worry about the list support in this PR?

It depends on what sort of support is missing.  If this PR is regressing lists from where they are now, then we need to either fix other things before this PR so it doesn't happen or fix that in this PR, depending on where the problem is.

Thank you for rebasing the PR.  I'll try to review it tomorrow (June 17), but if not, Apple has a company holiday Friday, so it won't be until Monday....

erjiaqing commented:
@woody-apple @vivien-apple Just added a workaround to make current code which depends on List still work. PTAL

vivien-apple reviewed (commented):
> > I would assume we don't need to worry about the list support in this PR?
> 
> It depends on what sort of support is missing. If this PR is regressing lists from where they are now, then we need to either fix other things before this PR so it doesn't happen or fix that in this PR, depending on where the problem is.
> 

This PR breaks lists in the sense that trying to read them always returns an empty list.

erjiaqing commented on src/app/InteractionModelEngine.h:
```diff
@@ -153,6 +153,17 @@ class InteractionModelEngine : public Messaging::ExchangeDelegate
 
+    /**
+     *  Retrieve a ReadClient that the SDK consumer can use to send do a read.  If the call succeeds, the consumer
+     *  is responsible for calling Shutdown() on the ReadClient once it's done using it.
```
This is a private member. So the only caller is IMEngine now.

erjiaqing commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -293,6 +293,98 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
+                Callback::Callback<DefaultFailureCallback>::FromCancelable(onFailureCallback);
+            cb->mCall(cb->mContext, EMBER_ZCL_STATUS_INVALID_VALUE);
+            return true;
+        }
```
Added workaround for List

erjiaqing commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -293,6 +293,98 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
+
 bool emberAfReadAttributesResponseCallback(ClusterId clusterId, uint8_t * message, uint16_t messageLen)
 {
     ChipLogProgress(Zcl, "ReadAttributesResponse:");
```
Would prefer another PR for this cleanup.

vivien-apple replied:
I would prefer to do it into this PR. We lack tests and this is not always obvious to see what is covered by the IM and what is not. Removing the dead code would help to see what is effectively removed and compare what is covered by IM and what's not.

erjiaqing replied:
Got it, I removed this function, as well as some other functions in process-global-messages..cpp
Added a seperate commit for such cleanup work for better review experience.

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -65,14 +65,20 @@ Engine::RetrieveClusterData(AttributeDataElement::Builder & aAttributeDataElemen
+    if (!ServerClusterAttributeExists(aClusterInfo.mEndpointId, aClusterInfo.mClusterId, aClusterInfo.mFieldId))
+    {
+        // Check if the required attribute does not exist, put a status with not exist code.
+        aAttributeDataElementBuilder.GetWriter()->Put(TLV::ContextTag(AttributeDataElement::kCsTag_Status), (int32_t) 1);
```
This line replaced using InteractionModel protocol code now.

erjiaqing commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -293,6 +293,98 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
+
+    // CHECK_MESSAGE_LENGTH(1);
+    // uint8_t status = chip::Encoding::Read8(message); // zclStatus
+    // LogStatus(status);
```
This code does not present on latest commit of this PR

erjiaqing commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +97,128 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+bool ServerClusterAttributeExists(chip::EndpointId aEndPointId, chip::ClusterId aClusterId, chip::AttributeId aAttributeId)
+{
+    // TODO: Currently, we are using cluster catalog from the ember library, this should be modified or replaced after several
+    // updates to Commands.
```
This function is removed. There will ba a issue to analysis & discuss the gap between ember attribute management and the spec for this issue.

erjiaqing commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +97,128 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+                  " ListIndex=%" PRIx8,
+                  aClusterInfo.mClusterId, ChipLogValueX64(aClusterInfo.mNodeId), aClusterInfo.mEndpointId, aClusterInfo.mFieldId,
+                  aClusterInfo.mListIndex);
+    uint8_t data[256];
```
Added a TODO for this.

Choose 256 since it can hold values as large as OCTET_STRING, which should meet our current need, and can hold some small lists.

vivien-apple replied:
Sadly, I have converted ember to use `uint16_t` from `uint8_t` for attributes size because that was needed by the operation credential cluster IIRC correctly. So I don't think it will meet our needs.

erjiaqing replied:
Now using a bufferwith size EMBER_AF_MAXIMUM_SEND_PAYLOAD_LENGTH, which is also the buffer size used in ember library.

vivien-apple replied:
I thought ember was using `ATTRIBUTE_LARGEST` which is a generated value ? https://github.com/project-chip/connectedhomeip/blob/30daae2b38d8c07c5797f073942113b774f8f299/src/app/util/attribute-table.cpp#L259

erjiaqing commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +97,128 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+    break;
+    case ZCL_OCTET_STRING_ATTRIBUTE_TYPE: // Octet string
+    case ZCL_CHAR_STRING_ATTRIBUTE_TYPE:  // Character string
+        aWriter.PutBytes(AttributeDataElement::kCsTag_Data, data + 1, data[0]);
```
Did some tiny refactor and pass a ByteSpan to writer so the code is clearer. Thanks

erjiaqing commented:
@woody-apple Thanks for report, however, I build and run bridge app locally but did not see the issue. Can you pull the latest checkout to see if the issue persists>

vivien-apple commented:
Seems like the `Darwin` CI is failing on. We need to figure out why: 
```
-[CHIPClustersTests testSendClusterTestClusterReadAttributeListOctetStringWithResponseHandler]
-[CHIPClustersTests testSendClusterTestClusterReadAttributeListStructOctetStringWithResponseHandler]
```

Trying the patch locally it also crashes as soon as you are trying to access an attribute that does not exists on the target cluster. For example trying to read `./out/debug/standalone/chip-tool onoff read off-wait-time 2` of the bridge app results into a crash on the accessory side instead of a `EMBER_ZCL_STATUS_UNSUPPORTED_ATTRIBUTE` response.
```Process 74676 stopped
* thread #4, queue = 'com.zigbee.chip.framework.controller.workqueue', stop reason = EXC_BAD_ACCESS (code=1, address=0x2)
    frame #0: 0x0000000100014dd0 chip-bridge-app`chip::app::ReadSingleClusterData(aClusterInfo=0x00000001000a8078, apWriter=0x000070000efbd780, apDataExists=0x0000000000000000) at ember-compatibility-functions.cpp:192:41
   189 	    record.attributeId      = aClusterInfo.mFieldId;
   190 	    record.manufacturerCode = 0;
   191 	    status                  = emAfReadOrWriteAttribute(&record, &metadata, data, sizeof(data), false); // write?
-> 192 	    dataType                = metadata->attributeType;
   193 	
   194 	    if (apDataExists != nullptr)
   195 	    {
Target 0: (chip-bridge-app) stopped.
(lldb) p metadata
(EmberAfAttributeMetadata *) $0 = nullptr
```

I'm all for moving forward with IM for the attributes bits. But we do lack some tests and for such a change, it is pretty clear that we are going to break some features.

I have started to add more tests for attributes into https://github.com/project-chip/connectedhomeip/pull/7711 but I missed the case of `EMBER_ZCL_STATUS_UNSUPPORTED_ATTRIBUTE` there.
There is also no tests for `List` at the moment there too, nor there is tests for `external` attributes. Those are just the one I can think of on top of my head, but there is likely others too.

What do you think of  first adding more tests to `TestCluster.yaml` in order to cover as much as possible, and then we can validate that this PR is not going to put us into trouble ?

erjiaqing commented:
> Seems like the `Darwin` CI is failing on. We need to figure out why:
> 
> ```
> -[CHIPClustersTests testSendClusterTestClusterReadAttributeListOctetStringWithResponseHandler]
> -[CHIPClustersTests testSendClusterTestClusterReadAttributeListStructOctetStringWithResponseHandler]
> ```
> 
> Trying the patch locally it also crashes as soon as you are trying to access an attribute that does not exists on the target cluster. For example trying to read `./out/debug/standalone/chip-tool onoff read off-wait-time 2` of the bridge app results into a crash on the accessory side instead of a `EMBER_ZCL_STATUS_UNSUPPORTED_ATTRIBUTE` response.
> 
> ```
> * thread #4, queue = 'com.zigbee.chip.framework.controller.workqueue', stop reason = EXC_BAD_ACCESS (code=1, address=0x2)
>     frame #0: 0x0000000100014dd0 chip-bridge-app`chip::app::ReadSingleClusterData(aClusterInfo=0x00000001000a8078, apWriter=0x000070000efbd780, apDataExists=0x0000000000000000) at ember-compatibility-functions.cpp:192:41
>    189 	    record.attributeId      = aClusterInfo.mFieldId;
>    190 	    record.manufacturerCode = 0;
>    191 	    status                  = emAfReadOrWriteAttribute(&record, &metadata, data, sizeof(data), false); // write?
> -> 192 	    dataType                = metadata->attributeType;
>    193 	
>    194 	    if (apDataExists != nullptr)
>    195 	    {
> Target 0: (chip-bridge-app) stopped.
> (lldb) p metadata
> (EmberAfAttributeMetadata *) $0 = nullptr
> ```
> 
> I'm all for moving forward with IM for the attributes bits. But we do lack some tests and for such a change, it is pretty clear that we are going to break some features.
> 
> I have started to add more tests for attributes into #7711 but I missed the case of `EMBER_ZCL_STATUS_UNSUPPORTED_ATTRIBUTE` there.

My fault, this is caused by `metadata`would be nullptr if the attribute not exists.

> There is also no tests for `List` at the moment there too, nor there is tests for `external` attributes. Those are just the one I can think of on top of my head, but there is likely others too.
> 

I'm fine with the lists, however, I don't think "external" attributes should block this PR, since this PR is on protocol / encoding side, and storage side is untouched.

> What do you think of first adding more tests to `TestCluster.yaml` in order to cover as much as possible, and then we can validate that this PR is not going to put us into trouble ?

I found `TestCluster ListOctetString` already contains tests for lists with valid values, I think we can move forward with this PR if the Darwin CI passed.

vivien-apple commented:
> > I'm all for moving forward with IM for the attributes bits. But we do lack some tests and for such a change, it is pretty clear that we are going to break some features.
> > I have started to add more tests for attributes into #7711 but I missed the case of `EMBER_ZCL_STATUS_UNSUPPORTED_ATTRIBUTE` there.
> 
> My fault, this is caused by `metadata`would be nullptr if the attribute not exists.
> 
> > There is also no tests for `List` at the moment there too, nor there is tests for `external` attributes. Those are just the one I can think of on top of my head, but there is likely others too.
> 
> I'm fine with the lists, however, I don't think "external" attributes should block this PR, since this PR is on protocol / encoding side, and storage side is untouched.

I'm agreeing on the fact that `external` should just work since you're using `emAfReadOrWriteAttribute`.

> 
> > What do you think of first adding more tests to `TestCluster.yaml` in order to cover as much as possible, and then we can validate that this PR is not going to put us into trouble ?
> 
> I found `TestCluster ListOctetString` already contains tests for lists with valid values, I think we can move forward with this PR if the Darwin CI passed.

It does not check any values at the moment, it just checks the return status (https://github.com/project-chip/connectedhomeip/blob/master/src/darwin/Framework/CHIPTests/CHIPClustersTests.m#L6955). 
This is why `List` was green in CI even if the previous version of this PR was not returning anything for List.

It seems like this PR is failing at some of the tests I have added into #7711. For some reasons it was working on Darwin, not sure what's going on here : https://github.com/project-chip/connectedhomeip/pull/7634/checks?check_run_id=2855386106 

I'm going to add a few more simple tests for attributes (mostly #7741). That would make it easier to validate the end results of this PR on the stack. Nothing fancy, just a bunch of basic attributes reading tests.

erjiaqing commented:
> > > What do you think of first adding more tests to `TestCluster.yaml` in order to cover as much as possible, and then we can validate that this PR is not going to put us into trouble ?
> > 
> > 
> > I found `TestCluster ListOctetString` already contains tests for lists with valid values, I think we can move forward with this PR if the Darwin CI passed.
> 
> It does not check any values at the moment, it just checks the return status (https://github.com/project-chip/connectedhomeip/blob/master/src/darwin/Framework/CHIPTests/CHIPClustersTests.m#L6955).
> This is why `List` was green in CI even if the previous version of this PR was not returning anything for List.
> 
> It seems like this PR is failing at some of the tests I have added into #7711. For some reasons it was working on Darwin, not sure what's going on here : https://github.com/project-chip/connectedhomeip/pull/7634/checks?check_run_id=2855386106

That is an issue caused by CRMP, for the higher protocol side, just temporary disabled it with a line of note, will enable it later.

> I'm going to add a few more simple tests for attributes (mostly #7741). That would make it easier to validate the end results of this PR on the stack. Nothing fancy, just a bunch of basic attributes reading tests.

Thanks, that will also give me more confidence when landing this PR. Just rebased the PR for resolving the CI issues.

vivien-apple commented on src/app/InteractionModelEngine.h:
```diff
@@ -204,7 +215,24 @@ void DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aC
+ *  @param[in]    apWriter          The TLVWriter for holding cluster data.
+ *  @param[out]   apDataExists      Tell whether the cluster data exist on server.
+ *
+ *  @retval  CHIP_NO_ERROR on success
```
```
Fetch attribute value and version info and write to the TLVWriter provided.
When the endpoint / cluster / attribute / event data specified by aClusterInfo does not exist, the corresponding interaction model error code will be put into the writer, CHIP_NO_ERROR will be returned and apDataExists will be set to false.
If the data exists on the server, the data (with tag kCsTag_Data) and the data version (with tag kCsTag_DataVersion) will be put into the TLVWriter and apDataExists will be set to true. TLVWriter error will be returned if any error occurred during encoding these values.
This function is implemented by CHIP as a part of cluster data storage & management.
The apWriter and apDataExists can be nullptr.```

vivien-apple commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +170,143 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+    {
+        // TODO: replace by enum after #7482
+        err = apWriter->Put(chip::TLV::ContextTag(AttributeDataElement::kCsTag_Status),
+                            Protocols::InteractionModel::ToUint16(Protocols::InteractionModel::ProtocolCode::UnsupportedAttribute));
```
Similarly to the previous comment, there can be a few different statuses. It does not necessarily mean `UnsupportedAttribute`

vivien-apple commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +170,143 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+    switch (BaseType(dataType))
+    {
+    case ZCL_NO_DATA_ATTRIBUTE_TYPE: // No data
+        SuccessOrExit(err = apWriter->PutNull(TLV::ContextTag(AttributeDataElement::kCsTag_Data)));
```
For this and the others places, please use `ReturnOnFailure`.

vivien-apple commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +170,143 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+    EmberAfStatus status;
+    record.endpoint         = aClusterInfo.mEndpointId;
+    record.clusterId        = aClusterInfo.mClusterId;
+    record.clusterMask      = CLUSTER_MASK_SERVER;
```
What about `Client` attributes ?

erjiaqing replied:
CHIP does not have client attributes.

Spec: 
The server cluster supports attribute data, events and cluster commands. The client cluster initiates interactions, including invokation of cluster commands.

vivien-apple commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -37,6 +47,72 @@ EmberAfClusterCommand imCompatibilityEmberAfCluster;
+    case ZCL_ARRAY_ATTRIBUTE_TYPE:
+        return ZCL_ARRAY_ATTRIBUTE_TYPE;
+    default:
+        return type;
```
Since there is a `default` case, is there any reason to not let the `case` that return the same type falls into `default` ? That would make `BaseType` a list of special cases with a default handler.

erjiaqing replied:
Fixed

vivien-apple commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -215,6 +215,7 @@ void LogStatus(uint8_t status)
         ChipLogProgress(Zcl, "  status: EMBER_ZCL_STATUS_SOFTWARE_FAILURE (0x%02x)", status);
         break;
     case EMBER_ZCL_STATUS_UNSUPPORTED_CLUSTER:
+
```
nit: Remove the extra line

vivien-apple commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -226,6 +227,98 @@ void LogStatus(uint8_t status)
+{
+    switch (status)
+    {
+    case 0x0: // Success
```
Can we directly use the constants from: https://github.com/project-chip/connectedhomeip/blob/1da1a92e5197ad49c1765436881ca6a690d188b6/src/protocols/interaction_model/Constants.h#L66 ?

erjiaqing replied:
Fixed

vivien-apple commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -293,6 +386,201 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
+            ChipLogProgress(Zcl, "  value: %s", value ? "True" : "False");
+
+            Callback::Callback<Int64uAttributeCallback> * cb =
+                Callback::Callback<Int64uAttributeCallback>::FromCancelable(onSuccessCallback);
```
This is a part that I`m not sure. `Callback::Callback<Int64sAttributeCallback>` and `Callback::Callback<Int64uAttributeCallback>` are nice because we do not have tons of different callbacks.

On the other hand the callbacks are strongly typed at the moment and they do not need to checks that the value fits into the type that is supposed to be the expected type from the spec.

Will it be possible to keep the conversion to the right type here, and remove the changes to the callbacks types from this PR ? That will still enable IM for Read, and we can examine/discuss the type change into an other PR without loosing time ?

erjiaqing replied:
I did reverted this change, and passed tests locally.
Although this is quite dangerious for calling unmatched functions, however, on the other hand, it seems to be safe for now.

erjiaqing commented:
@vivien-apple Thanks for review.

As for the error code of ember read attribute, there is no document of what error code will be returned of emberReadOrWriteAttribute, so I would suggest another PR for mapping ember error codes to im error codes as a follow up of #7482.

As for the function return types, I would suggest still using the code in this PR. (Also, the case of boolean is a typo, fixed).

Calling functions with unmatched type is dangerious, and I think the client is responsible to check if the server returned expected  attribute type.

I would suggest adding some kind of "expected type" for attribute read as a followup.

erjiaqing commented:
@vivien-apple PTAL

vivien-apple commented:
> @vivien-apple Thanks for review.
> 
> As for the error code of ember read attribute, there is no document of what error code will be returned of emberReadOrWriteAttribute, so I would suggest another PR for mapping ember error codes to im error codes as a follow up of #7482.

Please open a followup for this. I guess at some point there will be other errors (e.g No Read Access)
> 
> As for the function return types, I would suggest still using the code in this PR. (Also, the case of boolean is a typo, fixed).
> 
> Calling functions with unmatched type is dangerious, and I think the client is responsible to check if the server returned expected attribute type.

I was not suggesting to not check the type. I'm suggesting to check the type into `CHIPClientCallbacks.cpp` before casting and dispatching the response. It sounds a bit odd to me to let the client validate the type checking of a value that is constrained per spec.

erjiaqing commented:
> > @vivien-apple Thanks for review.
> > As for the error code of ember read attribute, there is no document of what error code will be returned of emberReadOrWriteAttribute, so I would suggest another PR for mapping ember error codes to im error codes as a follow up of #7482.
> 
> Please open a followup for this. I guess at some point there will be other errors (e.g No Read Access)

Issue #7817 opened

> 
> > As for the function return types, I would suggest still using the code in this PR. (Also, the case of boolean is a typo, fixed).
> > Calling functions with unmatched type is dangerious, and I think the client is responsible to check if the server returned expected attribute type.
> 
> I was not suggesting to not check the type. I'm suggesting to check the type into `CHIPClientCallbacks.cpp` before casting and dispatching the response. It sounds a bit odd to me to let the client validate the type checking of a value that is constrained per spec.

In this case, I would still suggest checking the types of values by client. Some reasons might be:
1. The response is valid per the spec -- the server puts a value that they thought it is correct per spec. --> The current failure callback cannot process such error status -- this is the case that the value is not expected.
2. The response does not contain type info -- since we are using TLV, the type info is no longer required, also it may limit the types the TLV can represent.
3. The client did not and have no way to tell what is the expected type they want.
4. Decoding & validating the values in the response result in huge function body even it is generated.

With the above discussion, I would still suggest only distinguish int / uint types, however, we can provide handy adapters as the middleware of the callback (as a follow up) in this case, the client still don't have to validate the values, and we can keep decoding & dispathing logic clean.

erjiaqing commented:
@vivien-apple @woody-apple PTAL

erjiaqing commented:
@vivien-apple @bzbarsky-apple @woody-apple PTAL

vivien-apple commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +157,141 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+        break;
+    }
+    default:
+        ChipLogError(DataManagement, "Attribute type %d not handled", static_cast<int>(dataType));
```
Please make the output the attribute code in hex. I'm pretty sure nobody is used to look at the attribute code in decimal.

vivien-apple commented on src/controller/python/templates/python-CHIPClusters-cpp.zapt:
```diff
@@ -37,9 +37,34 @@ void OnDefaultFailureResponse(void * /* context */, uint8_t status)
+        strValue += ' ';
+        strValue += std::to_string(value.data()[i]);
+    }
+    ChipLogProgress(Zcl, "  attributeValue: (span of length %zd) %s", value.size(), strValue.c_str());
```
That method won't print nice things for `OCTET_STRING` with non-printable characters.

erjiaqing replied:
This will print the string as a series of numbers.

erjiaqing commented on examples/pump-app/pump-common/gen/CHIPClientCallbacks.h:
```diff
@@ -43,6 +47,39 @@ typedef void (*Int32sAttributeCallback)(void * context, int32_t value);
+    }
+    else
+    {
+        ChipLogError(Zcl, "Failed to get value: %d", err);
```
Done

erjiaqing commented on examples/pump-app/pump-common/gen/CHIPClientCallbacks.cpp:
```diff
@@ -312,259 +485,88 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
+    {
+        err = data->GetDataPtr(val);
+        len = data->GetLength();
+    }
```
In TLV, we do not use 0xFFFF for invalid value, however, this do noticed me about the invalid string in ReadSingleClusterData implementation, thanks.

erjiaqing commented on src/app/ReadClient.cpp:
```diff
@@ -148,8 +150,10 @@ CHIP_ERROR ReadClient::SendReadRequest(NodeId aNodeId, Transport::AdminId aAdmin
+    // NOTE: Disable CRMP temporary, should be enabled later.
+    err = mpExchangeCtx->SendMessage(
+        Protocols::InteractionModel::MsgType::ReadRequest, std::move(msgBuf),
+        Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse).Set(Messaging::SendMessageFlags::kNoAutoRequestAck));
```
Added an issue number.

erjiaqing commented on src/app/util/CHIPDeviceCallbacksMgr.cpp:
```diff
@@ -72,9 +65,14 @@ CHIP_ERROR CHIPDeviceCallbacksMgr::AddResponseCallback(NodeId nodeId, uint8_t se
+    if (filter != nullptr)
+    {
+        ReturnErrorOnFailure(AddResponseFilter(info, filter));
+    }
```
Done

erjiaqing commented on src/app/util/CHIPDeviceCallbacksMgr.h:
```diff
@@ -31,11 +31,25 @@
 namespace chip {
 namespace app {
 
+constexpr size_t kTLVFilterPoolSize = 32;
```
I found there are another config called CHIPProjectConfig, I think it will be better if we use per project config instead of per platform config.

bzbarsky-apple commented on src/app/zap-templates/templates/app/CHIPClusters-src.zapt:
```diff
@@ -75,8 +76,30 @@ CHIP_ERROR {{asCamelCased name false}}Cluster::DiscoverAttributes(Callback::Canc
+    attributePath.mClusterId  = mClusterId;
+    attributePath.mFieldId    = {{asHex code 4}};
+    attributePath.mFlags.Set(chip::app::AttributePathParams::Flags::kFieldIdValid);
+    if (onSuccessCallback != nullptr || onFailureCallback != nullptr)
```
I don't follow this part.  The code in `CHIPClientCallbacks.zapt ` was assuming these callbacks are both non-null assuming there are any callbacks at all, as far as I can tell.  So here we either need to check both are non-null or assume they're both non-null.  Or we need to fix that code to deal with only one callback being null.

I guess the old code had the same problem, now that I've read that far?  Followup, please.

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -450,6 +450,7 @@ CHIP_ERROR DeviceController::GetDevice(NodeId deviceId, Device ** out_device)
             VerifyOrExit(err == CHIP_NO_ERROR, ReleaseDevice(device));
 
             device->Init(GetControllerDeviceInitParams(), mListenPort, mAdminId);
+            ReturnErrorOnFailure(device->LoadSecureSessionParametersIfNeeded());
```
This forces session establishment whenever `GetDevice` is called.  This does not seem right, and is the main thing that I don't think can be put off to a followup here.  Not least because session establishment is right now being converted to async, and making `GetDevice` async does not seem right.

The old code ended up doing this indirectly via `SendCommand` calling `SendMessage` on the device and then the device doing session setup.  For commands we do it via `Device::SendCommands`.

For this new setup, I think the best option is to have `{{asCamelCased parent.name false}}Cluster::ReadAttribute{{asCamelCased name false}}` call a method on the device to do a few things:

1. Set up the attribute path.  That is, we would pass in the things the device does not know (endpoint, cluster, field id) and have it do the work in only one place, so we don't have lots of duplicated code that all does the same thing.
2. Do the `LoadSecureSessionParameters`.  Then later we can update device code to do things async here.
3. Call SendReadRequest.

bzbarsky-apple reviewed (changes requested):
Some nits, mostly looks decent, but the GetDevice part really needs to be fixed...

erjiaqing commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +157,149 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+        break;
+    case ZCL_INT16U_ATTRIBUTE_TYPE: // Unsigned 16-bit integer
+    {
+        uint16_t uint16_data;
+        memcpy(&uint16_data, data, sizeof(uint16_data));
```
There might be some alignment issues on some embedded platforms. So memcpy is a more generalized solution. However, I just came up with a better solution that to use a template function to wrap this and the compiler can optimize it into reinterpret_cast equivalent.

bzbarsky-apple replied:
Alignment is a good point.  Please leave as memcpy for now and do the template function as a followup, to keep the scope down.

erjiaqing commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -23,20 +23,83 @@
 namespace Compatibility {
 namespace {
+constexpr uint32_t kTemporaryDataVersion  = 0;
+constexpr size_t kAttributeReadBufferSize = (ATTRIBUTE_LARGEST >= 8 ? ATTRIBUTE_LARGEST : 8);
```
I found there might be a attribute with size 3 on some applications.

bzbarsky-apple replied:
OK, but what's the issue with a size of 3?

erjiaqing replied:
When the server only contain a few small attributes, the `ATTRIBUTE_LARGEST` can be smaller than expected. In my case, it is 3, making the compiler complain that the buffer cannot hold a uint32.

bzbarsky-apple replied:
Followup is https://github.com/project-chip/connectedhomeip/issues/7937

erjiaqing commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -791,6 +760,86 @@ bool emberAfDiscoverCommandsReceivedResponseCallback(ClusterId clusterId, uint16
+    }
+    if (tlvError == CHIP_NO_ERROR)
+    {
+        tlvError   = reader.GetDataPtr(message);
```
The TAG here is not important, since the ReadClient should hand finished the parse work (including checking for the tag), and we only need to care about the value here.

yunhanw-google replied:
would do this in https://github.com/project-chip/connectedhomeip/issues/7937

erjiaqing commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -293,293 +466,89 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
-            }
+    if (CHIP_NO_ERROR != err)
+    {
+        if (onSuccessCallback == nullptr)
```
From my point of view, the functions (this and the others) in this file are used to handle & parse responses from server and then pass the values to the applications. So if callbacks are missing, this function does nothing infact.

bzbarsky-apple commented on src/app/ReadClient.h:
```diff
@@ -73,6 +73,9 @@ class ReadClient : public Messaging::ExchangeDelegate
                                size_t aAttributePathParamsListSize, EventNumber aEventNumber);
 
+    intptr_t GetAppIdentifier() const { return mAppIdentifier; }
+    Messaging::ExchangeContext * GetExchangeContext() const { return mpExchangeCtx; }
```
Tracked in https://github.com/project-chip/connectedhomeip/issues/7937

bzbarsky-apple commented on src/app/util/process-global-message.cpp:
```diff
@@ -155,82 +155,6 @@ bool emAfProcessGlobalCommand(EmberAfClusterCommand * cmd)
-
-        emberAfSendResponse();
-        return true;
-
```
Tracked in https://github.com/project-chip/connectedhomeip/issues/7937

bzbarsky-apple commented on src/app/InteractionModelEngine.h:
```diff
@@ -204,7 +216,26 @@ void DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aC
+ * The apWriter and apDataExists can be nullptr.
+ *
+ *  @param[in]    aClusterInfo      The cluster info object, for the path of cluster data.
+ *  @param[in]    apWriter          The TLVWriter for holding cluster data. Can be a nullptr if the caller does not care
```
This is still not making sense to me.  If the caller does not care about the value, what do they care about?  Just whether it exists?

But looking at the `ReadSingleClusterData` implemetations:

* The one in `TestCommandInteraction.cpp` leaves `apDataExists` unset if `apWriter` is null.
* The one in `TestReportingEngine.cpp`, same thing.
* The one in `chip_im_initiator.cpp`, same thing.
* The one in `chip_im_responder.cpp`, same thing.
* The one in `ember-compatibility-functions.cpp` seems to implement the "just set apDataExists" semantics, ok.

And more to the point: is there ever a caller that actually passes null for apWriter?  I don't obviously see one.

yunhanw-google replied:
https://github.com/project-chip/connectedhomeip/issues/7965

bzbarsky-apple commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -94,5 +157,149 @@ bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCom
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE: {
+        uint8_t * actualData = data + 2; // The pascal string contains 2 bytes length
+        uint16_t dataLength;
+        memcpy(&dataLength, data, sizeof(dataLength));
```
Tracked in https://github.com/project-chip/connectedhomeip/issues/7937

bzbarsky-apple commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -293,293 +466,89 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
+    Callback::Cancelable * onFailureCallback = nullptr;
+    app::TLVDataFilter tlvFilter             = nullptr;
+    NodeId sourceId                          = aPath.mNodeId;
+    // In CHIPClusters.cpp, we are using sequenceNumber as application identifier.
+    uint8_t sequenceNumber = static_cast<uint8_t>(apReadClient->GetAppIdentifier());
```
Tracked in https://github.com/project-chip/connectedhomeip/issues/7937

yunhanw-google commented on src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:
```diff
@@ -293,293 +432,88 @@ bool IMDefaultResponseCallback(const chip::app::Command * commandObj, EmberAfSta
-            }
+    if (CHIP_NO_ERROR != err)
+    {
+        if (onSuccessCallback == nullptr)
```
https://github.com/project-chip/connectedhomeip/issues/7966

woody-apple commented:
Approving given feedback. We'll have to follow up with require tests here, and future fixes.
